File: Data Structures & Algorithms/depth-of-binary-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    
    public int maxDepth(TreeNode root) {
        if(root == null)
            return 0;

        return depth(root, 1);
    }

    public int depth(TreeNode node, int depth) {
        // base
        if(node == null)
            return 0;

        // pre


        // recurse
        int leftDepth = depth(node.left, depth);
        int rightDepth = depth(node.right, depth);

        // post
        return 1 + Math.max(leftDepth, rightDepth);
    }

}

File: Data Structures & Algorithms/same-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        self.same = True

        def dfs(p, q):
            if not self.same: # early exist
                return

            # 1. base case
            if not p and not q: # both nodes are null so thats fine
                return
            
            # 2. recursive case

            # pre recursion
            if p and not q:
                self.same = False
                return

            if q and not p:
                self.same = False
                return

            if p and q and p.val != q.val:
                self.same = False

            # recurse
            dfs(p.left, q.left)
            dfs(p.right, q.right)

            # post recursion - no

        dfs(p, q)
        return self.same